README.md
[![Build Status](https://travis-ci.org/Ivana-/simple-datomic-cru.svg?branch=master)](https://travis-ci.org/Ivana-/simple-datomic-cru)

# Simple Datomic CRU

Allows to create, edit, see current state and history of orders entities in database.

List view with filters & history

![alt text](https://user-images.githubusercontent.com/10473034/61061608-386d7500-a405-11e9-9f15-fc60757c42dc.png "List view")

Edit form view

![alt text](https://user-images.githubusercontent.com/10473034/61061423-e6c4ea80-a404-11e9-911b-061c12da9b82.png "Edit order form")

## Join Datomic registration (if not yet)

<https://www.datomic.com/get-datomic.html> (Starter - free)

Current Datomic version, used in this project, is `0.9.5930` but you can change it in both `project.clj` and `Makefile`

## Installation

### Set environment variables

Create file `.env_secret` with following entry:

```
export DATOMIC_USERNAME=<Technical Contact from https://my.datomic.com/account>
export DATOMIC_PASSWORD=<Download Key from https://my.datomic.com/account>
export DATOMIC_LICENSE_KEY=<Your personal license key, sended you by email>
export DATOMIC_STORAGE_ADMIN_PASSWORD=<Storage admin password - any non-empty string without whitespaces>
export DATOMIC_STORAGE_DATOMIC_PASSWORD=<Storage user password - any non-empty string without whitespaces>
```

### And run install script

```
make install
```

## Development Mode

### Run server:

```
make start
```

or open any built-in REPL in your editor/IDE and evaluate `(-main)` from `core.clj`. Server will run on port 3000 by default.

### Run client:

```
lein clean
lein figwheel dev
```

Figwheel will automatically push cljs changes to the browser.

Wait a bit, then browse to [http://localhost:3449](http://localhost:3449).

## Testing

### Backend tests:

```
lein test
```

### Frontend tests:

```
lein doo chrome-headless test once
```

datomic-pro-0.9.5927/CHANGES.md
<!-- -*- mode: markdown ; mode: visual-line ; coding: utf-8 -*- -->

## Read This First

* This document covers all releases.  For a summary of critical
  release notices only, see
  https://docs.datomic.com/on-prem/release-notices.html.

* Releases with four version number components are bugfix-only releases
  based on their three-component parents, e.g. 0.8.4020.24 is a bugfix
  release for 0.8.4020. When upgrading, you should always choose the
  highest-numbered bugfix release in a family of releases.

* The Datomic Team recommends that you always test new releases in a
  staging environment first.

* The Datomic Team recommends that you always take a backup before
  adopting a new release.

## Changed in 0.9.5927

* New feature: Tuples.
  See https://docs.datomic.com/on-prem/schema.html#tuples.

* New feature: Attribute predicates.
  See https://docs.datomic.com/on-prem/schema.html#attribute-predicates.

* New feature: Entity specs.
  See https://docs.datomic.com/on-prem/schema.html#entity-specs.

* New feature: Return maps.
  See https://docs.datomic.com/on-prem/query.html#return-maps.
  
* New feature: Cassandra2 storage.
  See https://docs.datomic.com/on-prem/storage.html#cassandra2.

* Vulnerability Fix: Updated Commons Collections from 3.2.1 to 3.2.2.
  See https://issues.apache.org/jira/browse/COLLECTIONS-580.

* Fix: `sample` aggregate could hang query thread.

* Impoved performance of valcache cleanup.

* Updated REST server jetty dependencies from 9.3.7.v20160115 to
  9.4.15.v20190215.

* Updated ActiveMQ Artemis dependencies from 1.4.0 to 1.5.6.

## Changed in 0.9.5786

* Fix: re-enable memcached support which was inadvertently disabled in
  0.9.5783.

## Changed in 0.9.5783

* New feature: Valcache.
  See https://docs.datomic.com/on-prem/valcache.html.

* Fixed cache problem in peer-server where all `d/with` databases
  deriving from a common initial call to `d/with-db` had the same
  common value.

* Fixed bug introduced in 0.9.5703 that prevented using class static
  methods in query function expressions.

## Changed in 0.9.5703

* New feature: Classpath Functions.
  See https://docs.datomic.com/on-prem/database-functions.html#classpath-functions.

## Changed in 0.9.5697

* Release 0.9.5697 fixes two security vulnerabilities in Datomic
  On-Prem transactors running the free: or dev: storage protocol. See
  https://forum.datomic.com/t/important-security-update-0-9-5697/379.

* Enhancement: New security configuration options for free: and dev:
  transactors.

## Changed in 0.9.5661

* Upgrade: Updated Transactor AMIs for all regions to latest Amazon Linux. 
  Note that per AWS guidelines PV AMIs are no longer provided.

* Added support for eu-west-3 region

## Changed in 0.9.5656

* New feature: `:as` and `attr-with-opts` syntax in Pull API.
  See https://docs.datomic.com/pull.html.

## Changed in 0.9.5651

* Enhancement: Peer server allows arbitrary code in queries.

* Enhancement: Better error messages from peer server.

## Changed in 0.9.5561.62

* Bugfix: Transactors and peers now use disjoint ranges for tempid
  generation, so tempids generated in a transaction function cannot
  collide with tempids generated on the peer.

* Bugfix: Fixed a bug where starting a peer and transactor at about
  the same time could lead to a string tempid incorrectly unifying
  with another tempid in the same transaction.

## Changed in 0.9.5561.59

* Bugfix: Fixed bug that prevented healthcheck endpoint from starting
  on machines with more than 8 cores.

## Changed in 0.9.5561.56

* Bugfix: Stop printing sensitive Cassandra connection information in
  the log.

## Changed in 0.9.5561.54

* Bugfix: Fixed an HA performance bug where a transactor could
  continue to write a heartbeat during shutdown, causing up to a 30
  second delay before the standby transactor could take over.

* Enhancement: More efficient index use in transaction
  processing. This will improve write latency and throughput for some
  update-heavy transaction loads.

## Changed in 0.9.5561.50

* Bugfix: Release 0.9.5561.50 fixes a bug in the catalog that, in the
unlikely circumstance where one has deleted a database and restored it
from a backup without first having called `gc-deleted-dbs`, can cause
a subsequent `gc-deleted-dbs` to delete that (active) database.

* Bugfix: Prevent unbounded thread use by query pool.

* Upgrade: Peers and transactors now use version 1.11.82 of the AWS SDK.

* Upgrade: Updated the AWS regions and instances types available via
  the CloudFormation template.

* Enhancement: Better error message when unable to resolve an entity.

* Enhancement: Health check endpoints for transactors and peer servers.

## Changed in 0.9.5561

* Bugfix: Fixed a bug where peer server could ignore the `basis-t` of a
  client request, answering queries from the latest db value instead.

## Changed in 0.9.5554

* Bugfix: Fixed bug where queries with variables in attribute position
  could return tuples that differ only by Java representation, e.g. a
  tuple with Integer 1 and a 'different' tuple with Long 1.

* Bugfix: Fixed a bug where an ident that has been used to name more
  than one different entity over time could return an outdated entity
  id from e.g. `d/entid`.

* Improvement: Better performance creating memory databases.

## Changed in 0.9.5544

Bugfixes in new functionality introduced in 0.9.5530:

* clients can now query values produced by `with-db`.

* string tempids are now available in the `:tempids` key
  returned by e.g. `transact`.

* indexing jobs correctly handle the implicit use of
  `:db.install/attribute` and `:db.alter/attribute`.

## Changed in 0.9.5530

* New Feature: Clients and peer server. See
  https://docs.datomic.com/on-prem/clients-and-peers.html and
  https://docs.datomic.com/on-prem/peer-server.html

* New Feature: String tempids. See
  https://docs.datomic.com/on-prem/transactions.html#creating-temp-id

* Improvement: `db.install/attribute` and `db.alter/attribute` are now

test/ui/tt.clj
(ns ui.tt
  (:require
;    [matcho.core :as matcho]
   [re-frame.core :as rf]
;    [testing.fx :as fx]
;    [testing.gen :as gen]
;    [app.db :as db]
   [clojure.test :refer :all]
;   [ui.encounters.grid-model :as m]
   ;; [arrival-test-task.events]
   ;; [arrival-test-task.handler :as handler]
   ))

(deftest grid

  (is (= (count [1 2 3]) 3))

  ; (rf/dispatch [::model/test-fx "zazaza"])

  ; (def m (rf/subscribe [::model/get-value [:test]]))

  ; (def m (rf/subscribe [:get-values-by-paths {:test :test}]))


  ; (identity @m)

  ; (is (= 2 (count (:items @m))))

  ; (reset! re-frame.db/app-db {;;
  ;                             :date-from "sdrgsdfgsdfgsd"
  ;                             ;; :date-from "2019-07-14" ;; "2019-07-15"
  ;                             :date-to ""
  ;                             :text-search ""})

  ; (rf/dispatch [::model/load-order-list])
  
  ; (rf/dispatch [::model/db-test 345])
  
  ; ::model/load-order-list
  
  ; ; ::load-order-list
  ; ; ::save-order
  ; ; ::fetch-order
  ; ; ::fetch-order-history
  
  ; (def m (rf/subscribe [:get-values-by-paths {:order-list :order-list}]))

  ; (identity @m)
  ; (identity @re-frame.db/app-db)


  
  ; (reset! re-frame.db/app-db {})
  ; (rf/dispatch [::model/fetch-test 17592186045422])
  ; (rf/dispatch [::model/fetch-test 333])

  ; (identity @re-frame.db/app-db)

  ;;
  )

(comment
  (run-tests)
  ;;
  )

; (run-tests)
; (deftest grid

;   (gen/bronx)

;   (def encounters (gen/a! [:patient :appointment {:encounter 3}]
;                           {:encounter [{:period {:start "2019-02-23"}}
;                                        {:period {:start "2019-02-20"}}
;                                        {:period {:start "2019-02-23"}
;                                         :status "cancelled"}]}))

;   (is (= (count (map #(get-in % [:period])
;                      (fx/http [:get* "Encounter"])))
;          3))

;   (rf/dispatch [:encounters/grid :init {:params {:date "2019-02-22"}}])

;   (def m (rf/subscribe [:encounters/grid]))

;   (is (= 2 (count (:items @m))))

;   (matcho/assert
;    ^:matcho/strict
;    ["2019-02-23"
;     "2019-02-23"]
;    (map :start-time (:items @m)))


;   (rf/dispatch [:encounters/grid :init {:params {:date "2019-02-22"
;                                                  :status "cancelled"}}])

;   (def m (rf/subscribe [:encounters/grid]))

;   (is (= 1 (count (:items @m))))

;   (matcho/assert
;    ^:matcho/strict
;    ["cancelled"]
;    (map :status (:items @m)))

;   (rf/dispatch [:encounters/grid :init {:params {:date "2019-02-19"}}])

;   (def m (rf/subscribe [:encounters/grid]))

;   (is (= 1 (count (:items @m))))


;   (rf/dispatch [:encounters/grid :init {:params {:date "2019-02-10"}}])

;   (def m (rf/subscribe [:encounters/grid]))

;   (is (= 0 (count (:items @m)))))


project.clj
(defproject arrival-test-task "0.1.0-SNAPSHOT"
  :dependencies [[org.clojure/clojure "1.10.1"]
                 [org.clojure/clojurescript "1.10.520"]
                 [reagent "0.8.1"]
                 [re-frame "0.10.6"]
                 [garden "1.3.9"]
                 [ns-tracker "0.4.0"]
                 [compojure "1.6.1"]
                 [yogthos/config "1.1.2"]
                 [ring "1.7.1"]

                 ;; [com.datomic/client-pro "0.8.28"]
                 [com.datomic/datomic-pro "0.9.5930" :exclusions [com.google.guava/guava]]
                 [clj-commons/secretary "1.2.4"]

                 [day8.re-frame/test "0.1.5"]
                 [lein-doo "0.1.11"]
                 [devcards "0.2.6"]
                 ;;
                 ]

  :repositories {"my.datomic.com" {:url "https://my.datomic.com/repo"
                                   :username ~(System/getenv "DATOMIC_USERNAME")
                                   :password ~(System/getenv "DATOMIC_PASSWORD")}}

  :main arrival-test-task.core

  :plugins [[lein-cljsbuild "1.1.7"]
            [lein-garden "0.2.8"]
            [lein-environ "1.1.0"]
            [lein-doo "0.1.11"]]

  :min-lein-version "2.5.3"

  :source-paths ["src/clj" "src/cljs"]

  :clean-targets ^{:protect false} ["resources/public/js/compiled" "target"
                                    "resources/public/css"]

  :figwheel {:css-dirs ["resources/public/css"]
             ;; :ring-handler arrival_test_task.handler/dev-handler
             }

  :garden {:builds [{:id           "screen"
                     :source-paths ["src/clj"]
                     :stylesheet   arrival_test_task.css/screen
                     :compiler     {:output-to     "resources/public/css/screen.css"
                                    :pretty-print? true}}]}

  :profiles {:dev {:dependencies [[binaryage/devtools "0.9.10"]
                                  [re-frisk "0.5.4.1"]]

                   :plugins      [[lein-figwheel "0.5.18"]]

                  ; :source-paths ["test"]
                  ; :test-path "test"
                   
                   :extra-paths ["test"] ;; ["ui/srcs" "ui/test" "backend/test"]
                  ;  :extra-deps {com.cognitect/test-runner
                  ;               {:git/url "https://github.com/cognitect-labs/test-runner.git"
                  ;                :sha "3cb0a9daf1cb746259dc8309b218f9211ad3b33b"}
                                ; faker {:mvn/version "0.2.2"}
                                ; re-frame {:mvn/version "0.10.5"}
                                ; org.clojure/tools.cli {:mvn/version "0.4.1"}
                   
                                ; nrepl/nrepl {:mvn/version "0.6.0"}
                                ; refactor-nrepl {:mvn/version "2.4.0"}
                                ; cider/cider-nrepl {:mvn/version "0.22.0-beta4"}
                  ;              }
                   :env {:start-test-server? true}
                   ;;
                   }
             :prod {}
             :uberjar {:source-paths ["env/prod/clj"]
                       :omit-source  true
                       :main         arrival-test-task.core
                       :aot          [arrival-test-task.core]
                       :uberjar-name "arrival_test_task.jar"
                       :prep-tasks   ["compile" ["cljsbuild" "once" "min"]["garden" "once"]]}}

  ;; https://clojureverse.org/t/how-to-deal-with-development-code-in-clojurescript/613/2
  
  :cljsbuild {:builds [{:id           "dev"
                        :source-paths ["src/cljs"]
                        :figwheel     {:on-jsload "arrival_test_task.core/mount-root"}
                        :compiler     {:main                 arrival_test_task.core
                                       :output-to            "resources/public/js/compiled/app.js"
                                       :output-dir           "resources/public/js/compiled/out"
                                       :asset-path           "js/compiled/out"
                                       :source-map-timestamp true
                                       :preloads             [devtools.preload
                                                              re-frisk.preload]
                                       :external-config      {:devtools/config {:features-to-install :all}}}}

                       {:id           "min"
                        :source-paths ["src/cljs"]
                        :jar true
                        :compiler     {:main            arrival_test_task.core
                                       :output-to       "resources/public/js/compiled/app.js"
                                       :optimizations   :advanced
                                       :closure-defines {goog.DEBUG false}
                                       :pretty-print    false}}

                       {:id           "test"
                        :source-paths ["src/cljs" "test"] ;; ["src/cljs" "test/ui"];; ["src" "test"]
                        :compiler     {;; :preloads             [arrival-test-task.config]
                                       :main                 ui.test-runner ;; runners.doo
                                       :output-dir           "resources/public/js/compiled/test"
                                       :output-to            "resources/public/js/compiled/test.js"
                                       ;; :closure-defines currently does not have any effect
                                       ;; with :optimization :whitespace.
                                       ;; lein figwheel test - Error
                                       ;; the build :optimizations key is set to something other than :none
                                       :optimizations        :whitespace ;; :none ;; :advanced ;; :none ;; :whitespace
                                       ;; :closure-defines      {arrival-test-task.config/test? true}
                                       }}

                       #_{:id           "devcards-test"
                        :source-paths ["src/cljs" "test"] ;; ["src" "test"]
                        :figwheel     {:devcards true}
                        :compiler     {:main                 dev-cards.test-runner
                                       :optimizations        :none
                                       :asset-path           "js/compiled/devcards-test"
                                       :output-dir           "resources/public/js/compiled/devcards-test"
                                       :output-to            "resources/public/js/compiled/devcards-test.js"
                                       :source-map-timestamp true}}
                       ;;
                       ]})


src/clj/arrival_test_task/config.clj
(ns arrival-test-task.config
  (:require
   ;; [config.core :refer [env]]

   [clojure.string :as str]))

(def CONFIG nil)


(defn ne-string [x] (if ((every-pred string? not-empty) x) x ::error))


(def config-structure
  {:datomic-username ne-string
   :datomic-password ne-string
   :datomic-license-key ne-string
   :datomic-storage-admin-password ne-string
   :datomic-storage-datomic-password ne-string
   ;; :zazaza ne-string
   ;;
   })

(defn exit
  ([code template & args]
   (exit code (apply format template args)))

  ([code message]
   (let [out (if (zero? code) *out* *err*)]
     (binding [*out* out]
       (println message)))
   #_(System/exit code)))

(defn coerce-config [config]
  (try
    (let [{:keys [error result]}
          (reduce (fn [acc [k f]]
                    (let [x (get config k)
                          v (f x)]
                      (if (= ::error v)
                        (assoc-in acc [:error k] x)
                        (assoc-in acc [:result k] v))))
                  {} config-structure)]
      (if (empty? error)
        (do
          (println "Config successfully loaded")
          result)
        (exit 1 "Invalid config values: %s %s" \newline
              (str error)
              #_(-> error clojure.pprint/pprint with-out-str))))
    (catch Exception e
      (exit 1 "Wrong config values: %s" (ex-message e)))))


(defn remap-keys [m]
  (reduce (fn [acc [k v]] (assoc acc (-> k
                                         str/lower-case
                                         (str/replace #"_" "-")
                                         keyword) v)) {} m))

(defn parse-content [s type]
  (case type
    :env (->> s
              str/split-lines
              (remove str/blank?)
              (map #(let [[n v] (map str/trim (str/split % #"=" 2))
                          k (if (str/starts-with? n "export ") (str/trim (subs n 7)) n)]
                      [k v]))
              (into {})
              remap-keys)))

(defn read-config-file [filepath filetype]
  (try
    (-> filepath slurp (parse-content filetype))
    (catch Exception e nil)))

(defn set-config! [config] (alter-var-root (var CONFIG) (constantly config)))

(defn load-config! []
  (->> (merge
        (read-config-file ".env_secret" :env)
        (remap-keys (System/getenv)))
       coerce-config
       set-config!))


(comment
  (remap-env (System/getenv))
  (read-config-file ".env_secret" :env)
  ;; (keys env)
  (load-config!)
  CONFIG
  ;;
  )

src/clj/arrival_test_task/server.clj
(ns arrival-test-task.server
  (:require [arrival-test-task.handler :as handler]
            [config.core :refer [env]]
            [ring.adapter.jetty :refer [run-jetty]]))

(defonce server (atom nil))

(defn start-server! []
  (let [port (Integer/parseInt (or (env :port) "3000"))]
    (reset! server (cond-> {:main (run-jetty handler/dev-handler {:port port :join? false})}
                     (env :start-test-server?)
                     (assoc :test (run-jetty handler/test-handler {:port (inc port) :join? false}))))
    (println (str "Main server is running on port " port))
    (when (env :start-test-server?)
      (println (str "Test server is running on port " (inc port))))))

(defn stop-server! [] (doseq [x (vals @server)] (.stop x)))


(comment
  (start-server!)
  (identity @server)

  (do
    (stop-server!)
    (start-server!))
  ;;
  )

src/clj/arrival_test_task/handler.clj
(ns arrival-test-task.handler
  (:require [compojure.core :refer [GET POST OPTIONS defroutes]]
            [compojure.route :refer [resources]]
            [clojure.string :as str]
            [ring.util.response :refer [resource-response]]
            [arrival-test-task.datomic :as datomic]))


(defroutes routes
  
  (GET "/" [] (resource-response "index.html" {:root "public"}))
  
  (OPTIONS
    #_"/cors"
    "*" [] {:status 200
            :headers {"Access-Control-Allow-Origin" "*"
                      "Access-Control-Allow-Methods" "GET, POST, PATCH, PUT, DELETE"
                      "Access-Control-Allow-Headers" "Accept, Content-Type" ; "Origin, Content-Type, X-Auth-Token"
                      }})
  ;; (GET "/user/:id" [id] (str "<h1>Hello user " id "</h1>"))
  
  (POST "/clear-test-database" req (datomic/clear-test-database req))

  (GET  "/order"               req (datomic/order-list req))

  (GET  "/order/:id"           req (datomic/order-by-id req))
  
  (GET  "/order-history/:id"   req (datomic/order-history-by-id req))
  
  (POST "/order-save"          req (datomic/order-save req))

  (resources "/"))

;; (def dev-handler (-> #'routes wrap-reload))

(defn parse-query-params [query-string]
  (->> (str/split (str query-string) #"&")
       (map #(str/split % #"=" 2))
       (reduce (fn [acc [k v]] (assoc acc (keyword k) v)) {})))

; (defn wrap-test-db [handler]
;   (fn [req]
;     (handler (assoc req :test-db? true))))

(defn wrap-request [handler test-db?]
  (fn [{:keys [query-string body] :as req}]
    (handler (cond-> req
               query-string (assoc :query-params (parse-query-params query-string))
               body (update :body (fn [x] (try
                                            ;; clojure.edn/read-string
                                            (-> x slurp read-string)
                                            (catch Exception e x))))
               test-db? (assoc :test-db? true)))))

; (defn get-response-from-error
;   "Converts a thrown error into the response that should be
;    sent back to the client."
;   [error]
;   (let [type (get (ex-data error) :type)]
;     (case type
;       :not-found {:status 404 :body "Not found"}
;       {:status 500 :body "Internal server error"})))
; (defn wrap-error-handling
;   "Ring middleware that catches any thrown errors and sends an appropriate
;    response back to the client."
;   [handler]
;   (fn [req]
;     (try
;       (handler req)
;       (catch Exception e
;         (get-response-from-error e)))))


(defn wrap-response [handler]
  (fn [req]
    (let [res (handler req)]
      (-> res
          (update :body pr-str)
          (assoc-in [:headers "Content-type"] "text/clojure")
          (assoc-in [:headers "Access-Control-Allow-Origin"] "*")))))

(def dev-handler
  (-> routes
      (wrap-request false)
      wrap-response))

(def test-handler
  (-> routes
      (wrap-request true)
      wrap-response))


(comment
  dev-handler

  (defn handler [request]
    {:status 200
     :headers {"Content-Type" "text/html"}
     :body "Hello World"})

  (dev-handler 123)
  
  ; {:ssl-client-cert nil, 
;  :protocol "HTTP/1.1", 
;  :remote-addr "0:0:0:0:0:0:0:1", 
;  :params {:id "17592186045422"}, 
;  :route-params {:id "17592186045422"}, 
;  :headers {"origin" "http://localhost:3449", 
;            "host" "localhost:3000", 
;            "user-agent" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/75.0.3770.100 Safari/537.36", 
;            "content-type" "text/clojure", 
;            "referer" "http://localhost:3449/", 
;            "connection" "keep-alive", 
;            "pragma" "no-cache", 
;            "accept" "text/clojure", 
;            "accept-language" "ru-RU,ru;q=0.9,en-US;q=0.8,en;q=0.7", 
;            "accept-encoding" "gzip, deflate, br", 
;            "cache-control" "no-cache"}, 
;  :server-port 3000, 
;  :content-length nil, 
;  :compojure/route [:get "/order/:id"], 
;  :content-type "text/clojure", 
;  :character-encoding nil, 
;  :uri "/order/17592186045422", 
;  :server-name "localhost", 
;  :query-string nil, 
;  :body #object[org.eclipse.jetty.server.HttpInputOverHTTP 0x7045e554 "HttpInputOverHTTP@7045e554[c=0,q=0,[0]=null,s=STREAM]"], 
;  :scheme :http, 
;  :request-method :get
;  :test-db? true}

  ; ;; handler argument - incoming request
  ; {:ssl-client-cert nil, :protocol "HTTP/1.1", :remote-addr "0:0:0:0:0:0:0:1", 
  ;  :params {}, :route-params {}, 
  ;  :headers {"origin" "http://localhost:3449", "host" "localhost:3000", 
  ;            "user-agent" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/75.0.3770.100 Safari/537.36", 
  ;            "content-type" "text/clojure", "referer" "http://localhost:3449/", 
  ;            "connection" "keep-alive", "accept" "text/clojure", 
  ;            "accept-language" "ru-RU,ru;q=0.9,en-US;q=0.8,en;q=0.7", "accept-encoding" "gzip, deflate, br"}, 
  ;  :server-port 3000, :content-length nil, 
  ;  :compojure/route [:get "/order"], :content-type "text/clojure", 
  ;  :character-encoding nil, :uri "/order", :server-name "localhost", 
  ;  :query-string "date-from=2019-07-01", 
  ;  :body #object[org.eclipse.jetty.server.HttpInputOverHTTP 0x3459851e "HttpInputOverHTTP@3459851e[c=0,q=0,[0]=null,s=STREAM]"], 
  ;  :scheme :http, :request-method :get}
  
  ; {:ssl-client-cert nil, :protocol "HTTP/1.1", :remote-addr "0:0:0:0:0:0:0:1", 
  ;  :params {:id "17592186045422"}, :route-params {:id "17592186045422"}, 
  ;  :headers {"origin" "http://localhost:3449", "host" "localhost:3000", 
  ;            "user-agent" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/75.0.3770.100 Safari/537.36", 
  ;            "content-type" "text/clojure", "referer" "http://localhost:3449/", 
  ;            "connection" "keep-alive", "accept" "text/clojure", 
  ;            "accept-language" "ru-RU,ru;q=0.9,en-US;q=0.8,en;q=0.7", "accept-encoding" "gzip, deflate, br"}, 
  ;  :server-port 3000, :content-length nil, 
  ;  :compojure/route [:get "/order/:id"], :content-type "text/clojure", 
  ;  :character-encoding nil, :uri "/order/17592186045422", :server-name "localhost", 
  ;  :query-string nil, 
  ;  :body #object[org.eclipse.jetty.server.HttpInputOverHTTP 0x213bd797 "HttpInputOverHTTP@213bd797[c=0,q=0,[0]=null,s=STREAM]"], 
  ;  :scheme :http, :request-method :get}
  

  (dev-handler {:uri "/order"
                :query-string "date-from=2019-07-01"
                :request-method :get})
  ;;
  )
